chainer/dataset/tabular/_transform.py

```diff
@@ -0,0 +1,120 @@
+import six
+
+from chainer.dataset.tabular import tabular_dataset
+
+
+class _Transform(tabular_dataset.TabularDataset):
+
+    def __init__(self, dataset, keys, transform):
```

Comment:
Is it possible/reasonable to make `keys` automatically inferred like `mode`?

Reply:
No. If `transform` returns a `tuple`, we cannot know the name of keys. If `transform` returns a `dict`, we cannot know the order of keys.

Reply:
That makes sense. I still think the information of key names and the definition of transform are tightly connected, so it'd be good if the user only needs to specify `transform` and then the keys are automatically set. How about passing keys via an attribute/property of `transform`?

Reply:
> How about passing keys via an attribute/property of transform?

Do you mean the following?
```python
def transform(a, b, c):
    d = a + b
    e = b + c
    return d, e

transform.keys = ('d', 'e')
dataset.transform(transform)
```
I have two concerns.
1. Is is common to set an attribute to a function instance?
2. How can users specify `keys` for a lambda function?

Reply:
Perhaps, we should think about this PR (https://github.com/chainer/chainer/pull/7361) at the same time.

Reply:
I thought a similar approach as Trainer extensions, like:
```py3
class MyTransform:
    keys = 'd', 'e'
    def __call__(self, a, b, c):
        ...

# or

@tabular.make_transform(keys=('d', 'e'))  # it sets `keys` attribute
def my_transform(a, b, c):
    ...
```
For lambda functions and the cases when the user wants to override the `keys` attribute provided by the transform impl, we can keep optional `keys` argument to be specified at `transform` (this is also similar to `Trainer.extend`).

I haven't checked #7361 yet, so I will look through it. Thanks for pointing that out.

Reply:
If we provide a decorator `make_transform`, users can use it for lambda or overriding `keys`.
```python
tabular.make_transform(keys=('d', 'e'))(lambda a, b, c: ...)
```

Reply:
I think it will be confusing if there are more than one way to specify `keys`.

Reply:
`SimpleDataset` in https://github.com/chainer/chainer/pull/7276 also requires `key` and data of a column. This API also should be considered.

Reply:
I want to clarify my point of view. I'm mainly thinking about reusing the same transform function at multiple places (or by multiple users), like "let's use the same preprocessing as was done in the XXX project." In this case, there is no more than one way to specify `keys`; there are just two choices of keys for the user: the default keys, or custom (overriding) keys. The default keys exist for pushing "convention over configuration".

When the author of the transform just uses it immediately, there is no difference between the default keys and the overriding keys in effect, so you may feel there are multiple ways to do the same thing. Lambda function usage is an extreme case. I think just using the `keys` option of `transform` is much more natural than applying the decorator to the lambda, so I believe it won't be so confusing (but it's just my intuition).

As for `SimpleDataset.add_column` in #7276, who calls this method is always the same as who designs the dataset, so it's natural to always specify the key here. I don't think it affects the design choice for `transform`.

Reply:
> I want to clarify my point of view. I'm mainly thinking about reusing the same transform function at multiple places (or by multiple users), like "let's use the same preprocessing as was done in the XXX project."

Unlike trainer extensions, I think reusing of transform functions is not so common. So, `keys` attribute is not so useful. In most cases, the author of transform is same as the user.

Reply:
I see. If it is not so bothersome in most cases, then it's fine to leave it as is.

Reply:
If we find it is bothersome, we can add that functionality.

Reply:
OK, thanks for discussion!